source/common/ssl/context_manager_impl.cc

```diff
@@ -22,23 +58,77 @@ void ContextManagerImpl::releaseContext(Context* context) {
 
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
-  std::unique_lock<std::mutex> lock(contexts_lock_);
+ServerContextPtr ContextManagerImpl::createSslServerContext(
+    const std::string& listener_name, const std::vector<std::string>& server_names,
+    Stats::Scope& scope, ServerContextConfig& config, bool skip_context_update) {
+  ServerContextPtr context(new ServerContextImpl(*this, listener_name, server_names, scope, config,
+                                                 skip_context_update, runtime_));
+  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
```

Comment:
This non-trivial condition appears in multiple places.  Make it a helper?  
bool isWildcardServerName(const std::string& name)?

Reply:
Will do, but it feels like over-engineering to me.